src/preload.py

```diff
@@ -51,7 +51,7 @@
     "flash-boot",  # flasher images
 ]
 
-SUPERVISOR_REPOSITORY_RE = "^resin(playground)?/[a-z0-9]+-supervisor$"
+SUPERVISOR_REPOSITORY_RE = "^(?:(?:balena|resin|balenaplayground)??/)??(?:armel|rpi|armv7hf|aarch64|i386|amd64|i386-nlp)-supervisor$"
```

Comment:
I didn't get the inner `??`. It makes the expression match an empty repo such as `'/amd64-supervisor'`:
```
$ python3
>>> from re import match
>>> SUPERVISOR_REPOSITORY_RE = "^(?:(?:balena|resin|balenaplayground)??/)??(?:armel|rpi|armv7hf|aarch64|i386|amd64|i386-nlp)-supervisor$"

>>> not not match(SUPERVISOR_REPOSITORY_RE, '/amd64-supervisor')
True
```

```suggestion
SUPERVISOR_REPOSITORY_RE = "^((balena|resin|balenaplayground)/)?(armel|rpi|armv7hf|aarch64|i386|amd64|i386-nlp)-supervisor$"
```
 
Ideally we would not need to hardcode the list of possible balena architecture names here, because an architecture name could be added or renamed or aliased in the future. The SDK's `models.deviceType.getAllSupported` method could be used to get a list of "currently supported" architectures:
https://www.balena.io/docs/reference/sdk/node-sdk/#balena.models.deviceType.getAllSupported

However, "currently supported" architectures (as reported online by balena's backend) may not be exactly what we need, because a user may be trying to preload an old balenaOS image version for an architecture that we no longer support, and even downgrading the CLI would not help if the list of architectures came from balena's backend. The CLI's `balena devices supported --verbose --discontinued` command additionally uses the SDK's `models.config.getDeviceTypes` method to list discontinued device types, but the SDK has deprecated that method.
* https://www.balena.io/docs/reference/sdk/node-sdk/#balena.models.config.getDeviceTypes
* https://github.com/balena-io/balena-cli/blob/v12.44.20/lib/commands/devices/supported.ts#L89-L95

So, I don't know, maybe we go with a hardcoded list. If an architecture name is added or renamed or aliased in the future, some users would find that `balena preload` does not work as expected for their device type, they'd give up using balena or they'd contact support, GitHub issues would be created, `balena-preload` PR, `balena-cli` PR... (I understand that we don't have tests that exercise `balena preload` for all possible architectures, and even then the tests themselves should not hardcode the list of possible architectures.)

Not sure what the least bad option is. Maybe just revert to `[a-z0-9]+-supervisor` and match image names like `'not-supervisor'`? :-)

Reply:
hi @pdcastro, for me the worst option is erroneously matching something that is not the actual supervisor, as actions will be taken in that case. If we don't match for example a newer architecture we won't return a supervisor_version and won't take actions based on it.

The hardcoded list is based on https://github.com/balena-os/meta-balena/blob/51d2f658e973a999f9f346913157b7b4a6979900/meta-balena-common/recipes-containers/balena-supervisor/balena-supervisor.inc#L10.  So the OS is also hardcoding this, as it is seldom updated. 

I don't think the architectures returned by the API and the one used in supervisor names would actually match, so you would still need to keep a hardcoded table to match one with the other.

I'll update the PR with your improvement, I think that is actually the best we can do.